### PR TITLE
[WIP] Profiles 4.1.4.2 - Added a comparison for two Subjects

### PR DESCRIPTION
**Pull request by blen-desta**
## 4.1.4.2 - <Response> Usage

```
If multiple assertions are included, then each assertion's <Subject> element MUST refer 
to the same principal. It is allowable for the content of the <Subject> elements to differ 
(e.g. using different <NameID> or alternative <SubjectConfirmation> elements).
```

These `MUST`s in `Core - 3.3.4` are required to test the statement above

```
A <saml:Subject> element S1 strongly matches S2 if and only if the following two 
conditions both apply:

- If S2 includes an identifier element (<BaseID>, <NameID>,  or <EncryptedID>), 
then S1 MUST include an identical identifier element, but the element MAY be 
encrypted (or not) in either S1 or S2. In other words, the decrypted form of the 
identifier MUST be identical in S1 and S2. "Identical" means that the identifier 
element's content and attribute values MUST be the same. An encrypted 
identifier will be identical to the original according to this definition, once 
decrypted.

- If S2 includes one or more <saml:SubjectConfirmation> elements, then S1 
MUST include at least one <saml:SubjectConfirmation> element such that S1 
can be confirmed in the manner described by at least one
<saml:SubjectConfirmation> element in S2.

As an example of what is and is not permitted, S1 could contain a 
<saml:NameID> with a particular Format value, and S2 could contain a 
<saml:EncryptedID> element that is the result of encrypting S1's 
<saml:NameID> element. However, S1 and S2 cannot contain a 
<saml:NameID> element with different Format values and element 
content, even if the two identifiers are considered to refer to the
same principal.
```

### CODE REVIEW

File: ctk/common/src/main/kotlin/org/codice/compliance/utils/SubjectsComparison.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                }
+            }
+        }
+        return false
+    }
+
+    @SuppressWarnings("ReturnCount")
+    private fun idMatch(): Boolean {
+        val subject1Name = subject1.localName
+        val subject2Name = subject2.localName
```
Comment (brjeter) [inline ctk/common/src/main/kotlin/org/codice/compliance/utils/SubjectsComparison.kt]: ❗️ I'm fairly certain the `localName` property is just going to be "Subject" 100% of the time. I think what you mean to compare is the `localName` of the child element ending with "ID" for each one. I.e. you'd want to compare the `NameID` elements in this scenario
```
<Subject>
    <NameID>
        Name2
    </NameID>
</Subject>
<Subject>
    <NameID>
        Name1
    </NameID>
</Subject>
```

**Comment by brjeter** [comment]
Figured it out. 3.3.4 is technically under the "Assertion Query and Request Protocol". This similar section under "Authentication Request Protocol" is most directly relevant which explains the contradiction:
```
(3.4.1.4 Processing Rules)
If the <saml:Subject> element in the request is present, then the resulting assertions'
<saml:Subject> MUST strongly match the request <saml:Subject>, as described in 
Section 3.3.4, except that the identifier MAY be in a different format if specified by 
<NameIDPolicy>. In such a case, the identifier's physical content MAY be different, 
but it MUST refer to the same principal. [E75]If the identity provider cannot or will 
not produce assertions with a strongly matching subject, then it MUST return a 
<Response> with an error <Status>, and MAY return a second-level <StatusCode>
 that reflects the reason for the failure.
```
I don't know the details about it but the `<NameIdPolicy>` element is something on the request, so only the test would know about it. I think this is a verification that the test needs to call directly based on whether or not the test is expecting Subjects with identical NameID content or not.

File: ctk/common/src/main/kotlin/org/codice/compliance/verification/profile/SingleSignOnProfileVerifier.kt [hunk 2]
```diff
@@ -132,6 +135,20 @@ class SingleSignOnProfileVerifier(private val response: Node,
...
 
+    private fun verifySubjects(assertions: List<Node>) {
+        val allSubjects = assertions
+                .filter { it.children("Subject").isNotEmpty() }
+                .flatMap { it.children("Subject") }
+
+        val firstSubject = allSubjects.firstOrNull()
+
+        if (firstSubject != null
+                && allSubjects.any { !SubjectsComparison(it, firstSubject).subjectsMatch() })
```
[the remark this answers is not in the document]
Comment (blen-desta) [reply ctk/common/src/main/kotlin/org/codice/compliance/verification/profile/SingleSignOnProfileVerifier.kt]: From my understanding, I thought you could only have one method type. You're either bearer or holder-of-key. Not both. I may be wrong though. I'll double check

**Comment by blen-desta** [comment]
build now

**Comment by blen-desta** [comment]
build now

**Comment by blen-desta** [comment]
build now

**Comment by mackncheesiest** [comment]
build now

**Comment by blen-desta** [comment]
build now

**Comment by blen-desta** [comment]
build now

**Pull request closed without merging**